notes.md

```diff
@@ -0,0 +1,179 @@
+    Either of these 2 approaches may work:
+    - Only read socket when we finish processing a message. Pub-Sub app may queue up the frame for later processing, or it may immedieiately process it.
+    - Give FS2-Netty a queue to fill and only backpressure when queue is full (or above a high watermark, start reading when below a low watermark).
+- We want to send ws frames. 
+  - When sending any frame, we want to know if it was written to the connection. 
+  - When sending a close frame, then the frame itself should be sent, then the connection should be closed.
+  - We want to know if we're writing too fast and respond to backpressure (namely TCP backpressure).
+- We want to know why a websocket connection closed, what was the close code if any.
+
+WebSocket (extras that will save us time dev time if framework can do these):
+- We want to group connections, label an individual connection, write to a single connection, broadcast to group, close a single connection, and close all connections.
+- We want to know if connection is still alive.
+  - Close connection if it's unresponsive.
+- We want metrics for WS server
+  - Count of different frame types
+  - The size of data frames
+  - Count of different close codes
+    
+General Server configs:
+- We want to limit the number of TCP connections? To protect the server?
+    - Backpressure accepting connections (and reads) if memory is getting high?
+      E.g. Netty thread pool is queuing too many tasks in compute thread pool. Queue will hit OOM.
+
+## Implementation thoughts for above use cases
+Based on what currently exists and past discussions with additional modifications.
+
+## Package Structure
+- fs2-netty-core
+  - Netty
+    - wrappers
+    - ops
+- fs2-netty-server
+  - tcp
+  - udp
+- fs2-netty-client
+  - tcp
+  - udp
+- fs2-netty-all
+- fs2-netty-simple-http-server
+  - websockers
+
+### API
+We'll look at how intermediate library/module authors will use FS2-Netty to build protocols for above use-cases. 
+
+#### TCP Server
+The idea is that FS2-Netty users provide a custom Netty pipeline and probably some configs to `FS2TcpServer`, which
+builds the server resource that exposes a Stream of `Socket`'s. Then users can map over that to produce their desired
+api. `Socket` should be the bridge out of Netty, i.e. no more Netty further exposed (maybe this isn't practical,
+there's also more things Netty does besides sockets, so api will be refined). 
+```scala
+// similar to SocketHandler, a bridge out of Netty that exposes streams for I/O types and other TCP connection api methods 
+class NettyPipeline[I, O](handlers: List[NettyHandler])
+
+object FS2TcpServer {
+  def make[F[_], I, O](pipeline: NettyPipeline[I, O]): Resource[F, Fs2Stream[F, ServerSocket[F, I, O]]] = ???
+}
+
+object MySimpleHttpServer {
+
+  private trait HttpSocket[F[_]] {
+    def read(f: HttpRequest => F[HttpResponse]): F[Unit]
+  }
```

Comment:
I think it can actually be a bit more general than this! It would be something like `Socket[F, HttpRequest, HttpResponse]`. Which really just boils down to a pair of two things: `Stream[F, HttpRequest]` and `Pipe[F, HttpResponse, Unit]`. So to *write* to the socket, you need to supply a `Stream[F, HttpResponse]`, which you then feed to that pipe. To *read* from the socket, you simply consume from the stream of requests. Note that your `Stream[F, HttpResponse]` is probably a transformation of `Stream[F, HttpRequest]`. Thus, if you have a function `HttpRequest => F[HttpResponse]`, you can get the obvious semantic really easily:

```scala
val server: Stream[F, Socket[F, HttpRequest, HttpResponse]] = Network[F].server(host, port, httpPipeline) // or something

val handler: HttpRequest => F[HttpResponse] = ???

val handles = server map { (socket: Socket[F, HttpRequest, HttpResponse]) =>
  socket.reads
    .evalMap(handler)
    .through(socket.writes)
    .onError(e => logF("caught exception during request", e))
    .attempt
    .void
}

handles.parJoin(8196)     // limit to 8196 concurrent connections
```

Does that kind of make sense? Basically we're getting a *lot* from `Stream` for free here.

Reply:
Yes it does, I think.
So at a high level?
```
object ExampleHttpServer extends IOApp {
  private[this] val router = Kleisli[IO, FullHttpRequest, FullHttpResponse] {
    request => IO {
          new DefaultFullHttpResponse(
            HttpVersion.HTTP_1_1,
            HttpResponseStatus.NOT_FOUND
          )
        }
  }
  
  override def run(args: List[String]): IO[ExitCode] =
    HttpServer.start[IO](HttpServer.HttpConfigs(
      requestTimeoutPeriod = 500.milliseconds,
      HttpServer.HttpConfigs.Parsing.default
    ))
      .evalMap { httpClientConnections =>
        httpClientConnections
          .map(_.successfullyDecodedReads(router))
          .parJoin(65536)
          .compile
          .drain
      }
      .useForever
      .as(ExitCode.Success)
    
}
```
and under the hood, the overly specific `successfullyDecodedReads` method is the `handles` in your example. The `HttpServerConnection` in this PR (which I'll now call `HttpClientConnection`) will handle piping `Stream[F, HttpRequest]` through `Pipe[F, HttpResponse, Unit]` via `Kleisli[F, FullHttpRequest, FullHttpResponse]` transform.